tempodb/backend/gcs/gcs.go

```diff
@@ -30,33 +31,62 @@ type readerWriter struct {
 	hedgedBucket *storage.BucketHandle
 }
 
+var (
+	_ backend.RawReader             = (*readerWriter)(nil)
+	_ backend.RawWriter             = (*readerWriter)(nil)
+	_ backend.Compactor             = (*readerWriter)(nil)
+	_ backend.VersionedReaderWriter = (*readerWriter)(nil)
+)
+
 // NewNoConfirm gets the GCS backend without testing it
 func NewNoConfirm(cfg *Config) (backend.RawReader, backend.RawWriter, backend.Compactor, error) {
-	return internalNew(cfg, false)
+	rw, err := internalNew(cfg, false)
+	return rw, rw, rw, err
 }
 
 // New gets the GCS backend
 func New(cfg *Config) (backend.RawReader, backend.RawWriter, backend.Compactor, error) {
-	return internalNew(cfg, true)
+	rw, err := internalNew(cfg, true)
+	return rw, rw, rw, err
+}
+
+func NewVersionedReaderWriter(cfg *Config, confirmVersioning bool) (backend.VersionedReaderWriter, error) {
```

Comment:
I find the signature a bit redundant. We're instantiating a `backend.VersionedReaderWriter`, but if `confirmVersioning` is not `true`, then it really isn't versioned.

IMO we could keep `confirmVersioning` to the caller.

Reply:
The reason I pass it down to the gcs layer is because we need the client to lookup the bucket settings and verify versioning is enabled. So the caller can't do this (unless we expose more but I'd like to keep that limited).

I agree it's weird to create a VersionedReaderWriter but disable versioning. This boolean is more of a "check versioning is enabled, otherwise crash" setting. Disabling means "I don't care if versioning is actually enabled, let's just roll with it".
If versioning is not enabled the precondition headers just won't be checked but the operations will still work.

I added it because our e2e tests can't use versioning on gcs (unfortunately).

Alternatively we could also fall back to a regular RawReader and RawWriter but that would require a lot of annoying code because the method interface is different.

Reply:
I'm fine if the reason is keeping the rest of the code clean.